File: Semestru 4/MPP/lab/.net/lab3/lab3/Persistence/repo/implementation/EventDbRepository.cs
using Model.app.domain;
using Persistence.app.repo.@interface;
using Persistence.app.utils;
using log4net;
using System.Data.SQLite;

namespace Persistence.app.repo.implementation
{
	public class EventDbRepository : IEventRepository
	{
		private CommonUtils Utils;
		private static readonly ILog _logger = LogManager.GetLogger(typeof(EventDbRepository));

		public EventDbRepository()
		{
			_logger.Info("Initialized EventDbRepository.");
			this.Utils = new CommonUtils();
		}

		public Event Create(Event e)
		{
			_logger.Info($"Saving the event: {e.Name}, {e.MinAge} - {e.MaxAge}");
			try
			{
				_logger.Info("Succesfully connected to the db.");
				var connection = this.Utils.GetConnection();
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				using (var statement = new SQLiteCommand("insert into events (name, min_age, max_age) values (@name, @min_age, @max_age)", connection))
				{
					statement.Parameters.Clear();
					statement.Parameters.AddWithValue("@name", e.Name);
					statement.Parameters.AddWithValue("@min_age", e.MinAge);
					statement.Parameters.AddWithValue("@max_age", e.MaxAge);
					int rows = statement.ExecuteNonQuery();
					_logger.Info($"Affected rows: {rows}");
					return (Event)new Event(e.Name, e.MinAge, e.MaxAge).SetId((int)connection.LastInsertRowId);
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Event? Delete(int id)
		{
			throw new NotImplementedException();
		}

		public IEnumerable<Event> GetAll()
		{
			_logger.Info($"Searching for all events ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Event> events = new List<Event>();
				using (var query = new SQLiteCommand("select * from events",connection))
				{
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
							events.Add(
								(Event)new Event(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetInt32(reader.GetOrdinal("min_age")),
									reader.GetInt32(reader.GetOrdinal("max_age"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")))
						   );
						return events;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public IEnumerable<Event> GetAllByMaxAge(int maxAge)
		{
			_logger.Info($"Searching for all events ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Event> events = new List<Event>();
				using (var query = new SQLiteCommand("select * from events where max_age = @max_age", connection))
				{
					query.Parameters.Clear();
					query.Parameters.AddWithValue("@max_age", maxAge);
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
							events.Add(
								(Event)new Event(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetInt32(reader.GetOrdinal("min_age")),
									reader.GetInt32(reader.GetOrdinal("max_age"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")))
						   );
						return events;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public IEnumerable<Event> GetAllByMinAge(int minAge)
		{
			_logger.Info($"Searching for all events ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Event> events = new List<Event>();
				using (var query = new SQLiteCommand("select * from events where min_age = @min_age", connection))
				{
					query.Parameters.Clear();
					query.Parameters.AddWithValue("@min_age", minAge);
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
							events.Add(
								(Event)new Event(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetInt32(reader.GetOrdinal("min_age")),
									reader.GetInt32(reader.GetOrdinal("max_age"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")))
						   );
						return events;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Event? GetById(int id)
		{
			_logger.Info($"Searching for the event with id: {id}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				using (var query = new SQLiteCommand("select * from events where id = @id", connection))
				{
					query.Parameters.Clear();
					query.Parameters.AddWithValue("@id", id);
					using (var reader = query.ExecuteReader())
					{
						if (reader.Read())
							return (Event)new Event(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetInt32(reader.GetOrdinal("min_age")),
									reader.GetInt32(reader.GetOrdinal("max_age"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")));
						return null;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null;
		}

		public Event? Update(Event e)
		{
			_logger.Info($"Updating the event: {e.Name}, , {e.MinAge} - {e.MaxAge}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				var oldEvent = this.GetById(e.Id);
				if (oldEvent == null)
				{
					_logger.Info($"Didn't find any event with id = {e.Id}");
					return null!;
				}
				using (var statement = new SQLiteCommand("update events set name = @name, min_age = @min_age, max_age = @max_age where id = @id", connection))
				{
					statement.Parameters.Clear();
					statement.Parameters.AddWithValue("@name", e.Name);
					statement.Parameters.AddWithValue("@min_age", e.MinAge);
					statement.Parameters.AddWithValue("@max_age", e.MaxAge);
					statement.Parameters.AddWithValue("@id", e.Id);
					int rows = statement.ExecuteNonQuery();
					_logger.Info($"Affected rows: {rows}");
					return oldEvent;
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null;
		}
	}
}

File: Semestru 4/MPP/lab/java/lab3/Persistence/src/main/java/org/example/repository/db/EventDbRepository.java
package org.example.repository.db;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.domain.Event;
import org.example.repository.interf.IEventRepository;
import org.example.repository.utils.CommonUtils;


import java.sql.Connection;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class EventDbRepository implements IEventRepository {
    private CommonUtils Utils;
    private static final Logger logger = LogManager.getLogger();

    public EventDbRepository(Properties properties) {
        logger.info("Initializing ChildDbRepository with properties: {}", properties);
        this.Utils = new CommonUtils(properties);
    }

    @Override
    public Iterable<Event> GetAllByMinAge(int minAge) {
        logger.traceEntry("Getting all of the where min age = " + minAge);
        Connection connection = this.Utils.GetConnection();
        List<Event> events = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from events where min_age = ?")) {
            statement.setInt(1, minAge);
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next())
                    events.add(
                            (Event)new Event(
                                    resultSet.getString("name"),
                                    resultSet.getInt("min_age"),
                                    resultSet.getInt("max_age")
                            ).SetId(resultSet.getInt("id"))
                    );

            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return events;
    }

    @Override
    public Iterable<Event> GetAllByMaxAge(int maxAge) {
        logger.traceEntry("Getting all of the where max age = " + maxAge);
        Connection connection = this.Utils.GetConnection();
        List<Event> events = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from events where min_age = ?")) {
            statement.setInt(1, maxAge);
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next())
                    events.add(
                            (Event)new Event(
                                    resultSet.getString("name"),
                                    resultSet.getInt("min_age"),
                                    resultSet.getInt("max_age")
                            ).SetId(resultSet.getInt("id"))
                    );

            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return events;
    }

    @Override
    public Event FindById(Integer integer) {
        logger.traceEntry("Looking for event with id: {}", integer);
        Connection connection = this.Utils.GetConnection();
        Event event = null;
        try (var statement = connection.prepareStatement("select * from events where id = ?"))
        {
            statement.setInt(1, integer);
            try (var resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    event = (Event) new Event(resultSet.getString(2), resultSet.getInt(3), resultSet.getInt(4))
                            .SetId(resultSet.getInt(1));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return event;
    }

    @Override
    public Iterable<Event> FindAll() {
        logger.traceEntry("Getting all of the events.");
        Connection connection = this.Utils.GetConnection();
        List<Event> events = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from events")) {
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next())
                    events.add(
                            (Event)new Event(
                                    resultSet.getString("name"),
                                    resultSet.getInt("min_age"),
                                    resultSet.getInt("max_age")
                            ).SetId(resultSet.getInt("id"))
                    );

            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return events;
    }

    @Override
    public Event Save(Event entity) {
        logger.traceEntry("Saving event " + entity);
        Connection connection = this.Utils.GetConnection();
        try (var statement = connection.prepareStatement("insert into events (name, min_age, max_age) values (?, ?, ?)"))
        {
            statement.setString(1, entity.GetName());
            statement.setInt(2, entity.GetMinAge());
            statement.setInt(3, entity.GetMaxAge());
            int result = statement.executeUpdate();
            try (var generatedKeys = statement.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    int generatedId = generatedKeys.getInt(1);
                    return (Event) entity.SetId(generatedId);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error " + e);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Event Update(Event entity) {
        logger.traceEntry("Updating event " + entity);
        Connection connection = this.Utils.GetConnection();
        Event oldEvent = this.FindById(entity.GetId());
        try (var statement = connection.prepareStatement("update events set name = ?, min_age = ?, max_age = ? where id = ?"))
        {
            statement.setString(1, entity.GetName());
            statement.setInt(2, entity.GetMinAge());
            statement.setInt(3, entity.GetMaxAge());
            statement.setInt(4, entity.GetId());
            statement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error " + e);
        }
        logger.traceExit();
        return oldEvent;
    }

    @Override
    public Event Delete(Integer integer) {
        return null;
    }
}
